server/api_server/tools_routes.py

```diff
@@ -0,0 +1,686 @@
+
+    try:
+        # Run nmap -F for fast port scan
+        cmd = ["nmap", "-F", target]
+        result = subprocess.run(
+            cmd,
+            capture_output=True,
+            text=True,
+            check=True,
+            timeout=120
+        )
+
+        # Parse output for open ports
+        open_ports = []
+        for line in result.stdout.split('\n'):
+            if '/tcp' in line and 'open' in line:
+                parts = line.split('/')
+                port = parts[0].strip()
+                service = line.split()[2] if len(line.split()) > 2 else "unknown"
+                open_ports.append({"port": int(port), "service": service})
+
+        return jsonify({
+            "success": True,
+            "target": target,
+            "open_ports": open_ports,
+            "raw_output": result.stdout.strip().split('\n')
+        })
+
+    except subprocess.CalledProcessError as e:
+        return jsonify({"success": False, "error": "Port scan failed", "details": e.stderr.strip()}), 500
+    except Exception as e:
+        return jsonify({"error": str(e)}), 500
+
+
+@tools_bp.route('/get_network_topology', methods=['GET'])
+def get_network_topology():
+    """
+    Returns the "Parent/Child" relationships.
+    """
+    if not check_auth():
+        return jsonify({"error": "Unauthorized"}), 401
+
+    conn = get_temp_db_connection()
+    conn.row_factory = sqlite3.Row
+    cur = conn.cursor()
+
+    try:
+        cur.execute("SELECT devName, devMac, devParentMAC, devParentPort, devVendor FROM Devices")
+        rows = cur.fetchall()
+
+        nodes = []
+        links = []
+
+        for row in rows:
+            nodes.append({
+                "id": row['devMac'],
+                "name": row['devName'],
+                "vendor": row['devVendor']
+            })
+            if row['devParentMAC']:
+                links.append({
+                    "source": row['devParentMAC'],
+                    "target": row['devMac'],
+                    "port": row['devParentPort']
+                })
+
+        return jsonify({
+            "nodes": nodes,
+            "links": links
+        })
+    except Exception as e:
+        return jsonify({"error": str(e)}), 500
+    finally:
+        conn.close()
+
+
+@tools_bp.route('/get_recent_alerts', methods=['POST'])
+def get_recent_alerts():
+    """
+    Fetches the last N system alerts.
+    Arguments: hours (lookback period, default 24)
+    """
+    if not check_auth():
+        return jsonify({"error": "Unauthorized"}), 401
+
+    data = request.get_json()
+    hours = data.get('hours', 24)
+
+    conn = get_temp_db_connection()
+    conn.row_factory = sqlite3.Row
+    cur = conn.cursor()
+
+    try:
+        # Calculate cutoff time
+        cutoff = datetime.now() - timedelta(hours=int(hours))
+        cutoff_str = cutoff.strftime('%Y-%m-%d %H:%M:%S')
+
+        cur.execute("""
+            SELECT eve_DateTime, eve_EventType, eve_MAC, eve_IP, devName
```

Comment:
ideally use the NotificationInstance (would have to be extended as no matching method exists)